torch/optim/adagrad.py

```diff
@@ -103,13 +112,135 @@ def step(self, closure=None):
                     # record the step after step update
                     state_steps.append(state['step'])
 
-            F.adagrad(params_with_grad,
-                      grads,
-                      state_sums,
-                      state_steps,
-                      lr=group['lr'],
-                      weight_decay=group['weight_decay'],
-                      lr_decay=group['lr_decay'],
-                      eps=group['eps'])
+            adagrad(params_with_grad,
+                    grads,
+                    state_sums,
+                    state_steps,
+                    has_sparse_grad=has_sparse_grad,
+                    foreach=foreach,
+                    lr=group['lr'],
+                    weight_decay=group['weight_decay'],
+                    lr_decay=group['lr_decay'],
+                    eps=group['eps'])
 
         return loss
+
+
+def adagrad(params: List[Tensor],
+            grads: List[Tensor],
+            state_sums: List[Tensor],
+            state_steps: List[int],
+            has_sparse_grad: bool = False,
+            foreach: bool = False,
+            *,
+            lr: float,
+            weight_decay: float,
+            lr_decay: float,
+            eps: float):
+    r"""Functional API that performs Adagrad algorithm computation.
+
+    See :class:`~torch.optim.Adagrad` for details.
+    """
+
+    if foreach and not torch.jit.is_scripting() and (not has_sparse_grad):
+        func = _multi_tensor_adagrad
+    else:
+        func = _single_tensor_adagrad
+
+    func(params,
+         grads,
+         state_sums,
+         state_steps,
+         has_sparse_grad,
+         lr=lr,
+         weight_decay=weight_decay,
+         lr_decay=lr_decay,
+         eps=eps)
+
+def _make_sparse(grad, grad_indices, values):
+    size = grad.size()
+    if grad_indices.numel() == 0 or values.numel() == 0:
+        return torch.empty_like(grad)
+    return torch.sparse_coo_tensor(grad_indices, values, size)
+
+
+def _single_tensor_adagrad(params: List[Tensor],
+                           grads: List[Tensor],
+                           state_sums: List[Tensor],
+                           state_steps: List[int],
+                           has_sparse_grad: bool = False,
```

Comment:
added an unused kwarg with a default here `has_sparse_grad` so both functions take the same inputs and the above logic looks cleaner, is this ok?